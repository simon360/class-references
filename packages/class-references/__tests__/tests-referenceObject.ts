import {
  newObject,
  newObjectWithoutReferenceToToken,
  newObjectWithReferenceToClass
} from "../src/referenceObject";

describe("referenceObject", () => {
  const testClassName = "test-class-name";
  const extraTestClassName = "extra-test-class-name";

  describe("newObject()", () => {
    it("should match snapshot", () => {
      expect(newObject()).toMatchSnapshot();
    });
  });

  describe("newObjectWithReferenceToClass()", () => {
    it("should add a new class reference", () => {
      const obj = newObject();

      expect(
        newObjectWithReferenceToClass(obj, testClassName)
      ).toMatchSnapshot();
    });

    it("should add to an existing class reference", () => {
      let obj = newObject();

      obj = newObjectWithReferenceToClass(obj, testClassName);
      obj = newObjectWithReferenceToClass(obj, testClassName);

      expect(obj).toMatchSnapshot();
    });

    it("should add multiple class names", () => {
      let obj = newObject();

      obj = newObjectWithReferenceToClass(obj, testClassName);
      obj = newObjectWithReferenceToClass(obj, extraTestClassName);
      obj = newObjectWithReferenceToClass(obj, testClassName);

      expect(obj).toMatchSnapshot();
    });
  });

  describe("newObjectWithoutReferenceToId()", () => {
    it("should not change an empty object", () => {
      const obj = newObject();
      const objRemoved = newObjectWithoutReferenceToToken(obj, 0);

      expect(objRemoved).toEqual(obj);
    });

    it("should not remove a non-existent reference", () => {
      let obj = newObject();

      obj = newObjectWithReferenceToClass(obj, testClassName);

      const objRemoved = newObjectWithoutReferenceToToken(obj, 14);

      expect(objRemoved).toEqual(obj);
    });

    it("should not immediately remove an empty list", () => {
      let obj = newObject();

      obj = newObjectWithReferenceToClass(obj, testClassName);
      obj = newObjectWithoutReferenceToToken(obj, obj.lastToken);

      expect(obj).toMatchSnapshot();
    });

    it("should remove a received empty list", () => {
      let obj = newObject();

      obj = newObjectWithReferenceToClass(obj, testClassName);
      obj = newObjectWithoutReferenceToToken(obj, obj.lastToken);
      obj = newObjectWithoutReferenceToToken(obj, obj.lastToken); // Call again to prune empties

      expect(obj).toMatchSnapshot();
    });

    it("should leave unaffected class names alone", () => {
      let obj = newObject();

      const removableTokens = [];

      obj = newObjectWithReferenceToClass(obj, testClassName);
      removableTokens.push(obj.lastToken);
      obj = newObjectWithReferenceToClass(obj, testClassName);
      removableTokens.push(obj.lastToken);

      obj = newObjectWithReferenceToClass(obj, extraTestClassName);
      obj = newObjectWithReferenceToClass(obj, extraTestClassName);

      let objRemoved = { ...obj };

      for (const token of removableTokens) {
        objRemoved = newObjectWithoutReferenceToToken(objRemoved, token);
      }

      expect(obj.classes[extraTestClassName]).toEqual(
        objRemoved.classes[extraTestClassName]
      );
      expect(objRemoved.classes[testClassName]).toEqual([]);
    });
  });
});
