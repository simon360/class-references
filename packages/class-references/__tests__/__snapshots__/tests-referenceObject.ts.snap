// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`referenceObject newObject() should match snapshot 1`] = `
Object {
  "classes": Object {},
  "lastToken": 0,
}
`;

exports[`referenceObject newObjectWithReferenceToClass() should add a new class reference 1`] = `
Object {
  "classes": Object {
    "test-class-name": Array [
      1,
    ],
  },
  "lastToken": 1,
}
`;

exports[`referenceObject newObjectWithReferenceToClass() should add multiple class names 1`] = `
Object {
  "classes": Object {
    "extra-test-class-name": Array [
      2,
    ],
    "test-class-name": Array [
      1,
      3,
    ],
  },
  "lastToken": 3,
}
`;

exports[`referenceObject newObjectWithReferenceToClass() should add to an existing class reference 1`] = `
Object {
  "classes": Object {
    "test-class-name": Array [
      1,
      2,
    ],
  },
  "lastToken": 2,
}
`;

exports[`referenceObject newObjectWithoutReferenceToId() should not immediately remove an empty list 1`] = `
Object {
  "classes": Object {
    "test-class-name": Array [],
  },
  "lastToken": 1,
}
`;

exports[`referenceObject newObjectWithoutReferenceToId() should remove a received empty list 1`] = `
Object {
  "classes": Object {},
  "lastToken": 1,
}
`;
